nav2_pure_pursuit_controller/src/pure_pursuit_controller.cpp

```diff
@@ -99,93 +167,262 @@ void PurePursuitController::activate()
-    auto curvature = 2.0 * goal_pose.position.y /
-      (goal_pose.position.x * goal_pose.position.x + goal_pose.position.y * goal_pose.position.y);
-    linear_vel = desired_linear_vel_;
-    angular_vel = desired_linear_vel_ * curvature;
+  return false;
+}
+
+bool PurePursuitController::inCollision(const double & x, const double & y)
+{
+  using namespace nav2_costmap_2d;  // NOLINT
+
+  unsigned int mx, my;
+  costmap_->worldToMap(x, y, mx, my);
+
+  unsigned char cost = costmap_->getCost(mx, my);
+
+  if (costmap_ros_->getLayeredCostmap()->isTrackingUnknown()) {
+    return cost >= INSCRIBED_INFLATED_OBSTACLE && cost != NO_INFORMATION;
   } else {
-    linear_vel = 0.0;
-    angular_vel = max_angular_vel_;
+    return cost >= INSCRIBED_INFLATED_OBSTACLE;
   }
+}
 
-  // Create and publish a TwistStamped message with the desired velocity
-  geometry_msgs::msg::TwistStamped cmd_vel;
-  cmd_vel.header.frame_id = pose.header.frame_id;
-  cmd_vel.header.stamp = clock_->now();
-  cmd_vel.twist.linear.x = linear_vel;
-  cmd_vel.twist.angular.z = max(
-    -1.0 * abs(max_angular_vel_), min(
-      angular_vel, abs(
-        max_angular_vel_)));
+void PurePursuitController::applyConstraints(
+  double & linear_vel, double & angular_vel,
+  const double & dist_error, const double & lookahead_dist,
+  const geometry_msgs::msg::Twist & curr_speed)
+{
+  // if the actual lookahead distance is shorter than requested, that means we're at the
+  // end of the path. We'll scale linear velocity by error to slow to a smooth stop
+  if (approach_vel_scaling_ && dist_error > 2.0 * costmap_->getResolution()) {
+    double velocity_scaling = 1.0 - (dist_error / lookahead_dist);
+    if (velocity_scaling < min_approach_vel_scaling_) {
+      velocity_scaling = min_approach_vel_scaling_;
+    }
+    linear_vel = linear_vel * velocity_scaling;
+  }
 
-  return cmd_vel;
+  // make sure linear velocities are kinematically feasible, v = v0 + a*t
+  double & dt = control_duration_;
+  const double max_feasible_linear_speed = curr_speed.linear.x + max_linear_accel_ * dt;
+  const double min_feasible_linear_speed = curr_speed.linear.x - max_linear_decel_ * dt;
+  linear_vel = std::clamp(linear_vel, min_feasible_linear_speed, max_feasible_linear_speed);
+
+  // Note(stevemacenski): for the moment, no smoothing on angular velocities, we find the existing
+  // commands are very smooth and we don't want to artifically reduce them if the hardware
+  // can handle it. Plus deviating from the commands here can be collision-inducing if users
+  // don't properly set these values, so hiding that complexity from them that would likely
+  // become the #1 cause of issues.
+  // const double max_feasible_angular_speed = curr_speed.angular.z + angular_accel_ * dt;
+  // const double min_feasible_angular_speed = curr_speed.angular.z - angular_accel_ * dt;
+  // angular_vel = std::clamp(angular_vel, min_feasible_angular_speed, max_feasible_angular_speed);
```

Comment:
From experience, I've also noticed that smoothing of angular velocities is very bad for tracking performance. It's good to leave it out 👍

Reply:
Yeah, I was a little surprised to see this cause such large issues. I left it in in case someone came across it later and needed it. I pumped up the acceleration numbers too to make it well within the acceptable range but for some reason it still resulted in serious path deviations in confined spaces. I'll go back before merging this into nav2 master to make sure I fully understand why and if there's a fix, but it's probably not coming back.